src/components/tagging/tagging.js

```diff
@@ -0,0 +1,250 @@
+import { localize } from 'common/_locales/locales'
+import { Chips } from 'components/chips/chips'
+
+const taggingWrapper = css`
+  padding: 5px 0 0;
+  position: relative;
+`
+const taggingPlaceholder = css`
+  color: #999;
+  left: 29px;
+  position: absolute;
+  top: 10px;
+`
+const taggingWell = css`
+  background: #fff;
+  background-image: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100" fill="%23999999"><path d="M83.38 54.6L43.8 17.02H17.41V43.4l39.58 39.58L83.38 54.6zM30.95 36.81c-3.45 0-6.25-2.67-6.25-5.96s2.8-5.97 6.25-5.97 6.25 2.67 6.25 5.97c0 3.29-2.8 5.96-6.25 5.96z"/></svg>');
+  background-position: 4px 7px;
+  background-repeat: no-repeat;
+  background-size: 22px;
+  border: 1px solid #ddd;
+  border-radius: 3px;
+  box-sizing: border-box;
+  font-size: 13px;
+  line-height: 16px;
+  margin: 0;
+  padding: 4px 10px;
+  position: relative;
+  text-align: left;
+`
+
+const taggingTypeaheadWrapper = css`
+  position: relative;
+`
+
+const taggingTypeaheadItem = css`
+  cursor: pointer;
+  display: block;
+  padding: 2px 8px;
+
+  &.active {
+    background-color: #50bcb6;
+    color: #fff;
+  }
+  &:hover {
+    background-color: #50bcb6;
+    color: #fff;
+  }
+`
+const taggingTypeaheadList = css`
+  background: #fff;
+  border: 1px solid #ddd;
+  border-radius: 0 0 3px 3px;
+  border-top: none;
+  box-shadow: 0 4px 4px rgba(0, 0, 0, 0.2);
+  box-sizing: border-box;
+  display: block;
+  left: 0;
+  list-style-type: none;
+  margin: 0;
+  max-height: 8.8em;
+  overflow-x: hidden;
+  overflow-y: auto;
+  padding: 2px 0;
+  position: absolute;
+  top: 0;
+  width: 100%;
+`
+
+export const Tagging = ({
+  usedTags,
+  markedTags,
+  suggestedTags,
+  storedTags,
+  addTag,
+  activateTag,
+  deactivateTag,
+  deactivateTags,
+  removeTag,
+  removeTags,
+  closePanel
+}) => {
+  const hasTags = usedTags && usedTags.length
+
+  const [placeholder, setPlaceholder] = useState(!hasTags)
+  const [inputvalue, setInputValue] = useState('')
+  const inputRef = useRef(null)
+
+  /* Input Management
+    –––––––––––––––––––––––––––––––––––––––––––––––––– */
+  const setFocus = () => {
+    setPlaceholder(false)
+  }
+  const setBlur = () => {
+    const status = inputvalue.length || hasTags
+    setPlaceholder(!status)
+  }
+
+  /* Tag Management
+    –––––––––––––––––––––––––––––––––––––––––––––––––– */
+  const addTagAction = (value) => {
+    if (value === '') return
+    if (usedTags.indexOf(value) >= 0) return
+    addTag({ value })
+    setPlaceholder(false)
+    setInputValue('')
+    inputRef.current.focus()
+  }
+
+  /* Active/Inactive Tagging
+    –––––––––––––––––––––––––––––––––––––––––––––––––– */
+  const makeTagActive = (tag) => {
+    return activateTag({ tag })
+  }
+
+  const makeTagInactive = (tag) => {
+    return deactivateTag({ tag })
+  }
+
+  const makeTagsInactive = (blur) => {
+    if (!markedTags.length) return blur ? inputRef.current.blur() : null
```

Comment:
What is `blur` in this context?

Reply:
Blur is when the input box loses focus.

Reply:
Specifically, we pass this function to the input. If someone presses the ESCAPE key we call this function and pass in a `bool` for `blur` to make the input lose focus and "deactive' the tags in the input (which means the X disappears).